fdbserver/include/fdbserver/TenantEntryCache.actor.h

```diff
@@ -0,0 +1,336 @@
+		}
+	}
+
+	ACTOR static Future<Optional<TenantEntryCachePayload<T>>> getByIdImpl(TenantEntryCache<T>* cache,
+	                                                                      int64_t tenantId) {
+		Optional<TenantEntryCachePayload<T>> ret = cache->lookupById(tenantId);
+		if (ret.present()) {
+			return ret;
+		}
+
+		TraceEvent(SevInfo, "TenantEntryCache_GetByIdRefresh").detail("TenantId", tenantId);
+
+		// Entry not found. Refresh cacheEntries by scanning underling KeyRange.
+		// TODO: Cache will implement a "KeyRange" watch, monitoring notification when a new entry gets added or any
+		// existing entry gets updated within the KeyRange of interest. Hence, misses would be very rare
+		wait(refreshImpl(cache, TenantEntryCacheRefreshReason::CACHE_MISS));
+
+		return cache->lookupById(tenantId);
+	}
+
+	ACTOR static Future<Optional<TenantEntryCachePayload<T>>> getByNameImpl(TenantEntryCache<T>* cache,
+	                                                                        TenantName name) {
+		Optional<TenantEntryCachePayload<T>> ret = cache->lookupByName(name);
+		if (ret.present()) {
+			return ret;
+		}
+
+		TraceEvent(SevInfo, "TenantEntryCache_GetByNameRefresh").detail("TenantName", name.contents().toString());
+
+		// Entry not found. Refresh cacheEntries by scanning underling KeyRange.
+		// TODO: Cache will implement a "KeyRange" watch, monitoring notification when a new entry gets added or any
+		// existing entry gets updated within the KeyRange of interest. Hence, misses would be very rare
+		wait(refreshImpl(cache, TenantEntryCacheRefreshReason::CACHE_MISS));
+
+		return cache->lookupByName(name);
+	}
+
+	ACTOR static Future<Optional<TenantEntryCachePayload<T>>> getByIdImpl(TenantEntryCache<T>* cache, TenantName name) {
+		Optional<TenantEntryCachePayload<T>> ret = cache->lookupByName(name);
+		if (ret.present()) {
+			return ret;
+		}
+
+		TraceEvent(SevInfo, "TenantEntryCache_GetByNameRefresh").detail("TenantName", name.contents().toString());
+
+		// Entry not found. Refresh cacheEntries by scanning underling KeyRange.
+		// TODO: Cache will implement a "KeyRange" watch, monitoring notification when a new entry gets added or any
+		// existing entry gets updated within the KeyRange of interest. Hence, misses would be very rare
+		wait(refreshImpl(cache, TenantEntryCacheRefreshReason::CACHE_MISS));
+
+		return cache->lookupByName(name);
+	}
+
+	void incGen() { curGen++; }
+	Future<Void> waitForSweepTrigger() { return sweepTrigger.onTrigger(); }
+	void triggerSweeper() { sweepTrigger.trigger(); }
+
+	void sweep() {
+		TraceEvent(SevInfo, "TenantEntryCache_SweepStart", uid).log();
+
+		// Sweep cache and clean up entries older than current generation
+		TenantNameEntryPairVec entriesToRemove;
+		for (auto& e : mapByTenantId) {
+			ASSERT(e.value.gen <= curGen);
+
+			if (e.value.gen != curGen) {
+				entriesToRemove.emplace_back(std::make_pair(e.value.name, e.value.entry));
+			}
+		}
+
+		for (auto& r : entriesToRemove) {
```

Comment:
You could do this a little more cpu efficiently if you just did the mapByTenantId erase with the iterator as you were scanning the map, and just buffered the names to erase later (or also erased the names inline in one pass, then you wouldn't have to build a vector at all)

Reply:
Absolutely! :)
Thanks for pointing out.

Reply:
Looking more into `IndexedSet::erase` implementation, I think at the moment it doesn't advance the iterator the erase operation?

```
	// Remove the data item, if any, which is equal to key
	template <class Key>
	void erase(const Key& key) {
		erase(find(key));
	}
```

Reply:
Yeah, when I wrote something like this earlier, this was the reason to gather the entries to be removed in a separate vector.